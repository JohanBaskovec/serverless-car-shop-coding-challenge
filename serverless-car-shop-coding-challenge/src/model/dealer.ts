import mongoose, { Schema } from 'mongoose';
import { type DealerDocument, type DealerModel, type DealerSchema } from '../generated/mongoose.gen.js';

export const dealerSchema: DealerSchema = new Schema<DealerDocument>({
    name: {
        type: String,
        required: true
    },
    vehicles: [{
        type: Schema.Types.ObjectId,
        ref: 'Vehicle',
        required: true
    }]
}, {
    timestamps: {
        createdAt: 'creationDate',
        updatedAt: 'lastUpdateDate'
    }
});

export const Dealer: DealerModel = mongoose.model<DealerDocument, DealerModel>('Dealer', dealerSchema);
