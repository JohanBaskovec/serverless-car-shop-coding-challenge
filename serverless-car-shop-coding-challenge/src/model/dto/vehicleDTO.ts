import { type components } from '../../generated/openapi.js';

export type CreateVehicleDTO = components['schemas']['CreateVehicleDTO'];
export type UpdateVehicleDTO = components['schemas']['UpdateVehicleDTO'];
