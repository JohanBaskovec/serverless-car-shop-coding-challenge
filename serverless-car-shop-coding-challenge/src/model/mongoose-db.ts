import { connect, type Mongoose } from 'mongoose';

export async function connectToMongo(): Promise<Mongoose> {
    if (process.env.DB_URL == null) {
        throw new Error('process.env.DB_URL must be defined.');
    }
    return await connect(process.env.DB_URL);
}
