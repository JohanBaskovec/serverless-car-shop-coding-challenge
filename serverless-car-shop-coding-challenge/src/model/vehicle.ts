import mongoose, { Schema } from 'mongoose';
import { type VehicleDocument, type VehicleModel, type VehicleSchema } from '../generated/mongoose.gen';
import { type components } from '../generated/openapi.js';

export type FuelType = components['schemas']['FuelType'];

export type VehicleType = components['schemas']['VehicleType'];

export const vehicleSchema: VehicleSchema = new Schema<VehicleDocument>({
    make: {
        type: String,
        required: true
    },
    // naming the property "model" makes it impossible to mock the module
    vehicleModel: {
        type: String,
        required: true
    },
    transmission: {
        type: String,
        required: true,
        enum: [
            'manual',
            'automatic',
            'semiAutomatic'
        ]
    },
    fuelType: {
        type: String,
        required: true,
        enum: [
            'petrol',
            'diesel',
            'eletric',
            'lpg',
            'hybrid'
        ] as FuelType
    },
    mileage: {
        type: Number,
        required: true
    },
    vehicleType: {
        type: String,
        required: true,
        enum: [
            'cabriolet',
            'coupe',
            'estateCar',
            'suv',
            'saloon',
            'van',
            'smallCar',
            'other'
        ] as VehicleType
    },
    color: {
        type: String,
        required: true
    },
    dealer: {
        type: Schema.Types.ObjectId,
        ref: 'Dealer',
        required: true
    }
}, {
    timestamps: {
        createdAt: 'creationDate',
        updatedAt: 'lastUpdateDate'
    }
});

export const Vehicle: VehicleModel = mongoose.model<VehicleDocument, VehicleModel>('Vehicle', vehicleSchema);
