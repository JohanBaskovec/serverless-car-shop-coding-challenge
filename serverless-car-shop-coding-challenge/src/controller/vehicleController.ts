import { type CreateVehicleDTO, type UpdateVehicleDTO } from '../model/dto/vehicleDTO.js';
import { CrudController } from './crudController.js';
import { type VehicleDocument } from '../generated/mongoose.gen.js';

export class VehicleController extends CrudController<VehicleDocument, CreateVehicleDTO, UpdateVehicleDTO> {
}
