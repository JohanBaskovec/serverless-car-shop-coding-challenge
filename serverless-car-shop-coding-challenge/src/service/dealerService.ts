import { Dealer } from '../model/dealer.js';
import { type CreateDealerDTO, type UpdateDealerDTO } from '../model/dto/dealerDTO.js';
import { ConstraintError, EntityDoesNotExistError, MissingKeyInJSONBodyError } from '../error/errors.js';
import mongoose from 'mongoose';
import { type VehicleService } from './vehicleService.js';
import { type DealerDocument, type VehicleDocument } from '../generated/mongoose.gen.js';
import { type CrudService } from './crudService.js';
import { type APIGatewayProxyEventQueryStringParameters } from 'aws-lambda/trigger/api-gateway-proxy.js';
import { type DeleteResponse } from '../model/response.js';

export class DealerService implements CrudService<DealerDocument, CreateDealerDTO, UpdateDealerDTO> {
    private _vehicleService: VehicleService;

    get vehicleService(): VehicleService {
        return this._vehicleService;
    }

    set vehicleService(value: VehicleService) {
        this._vehicleService = value;
    }

    async getAll(queryStringParameters: APIGatewayProxyEventQueryStringParameters | null = null): Promise<DealerDocument[]> {
        return await Dealer.find({}).exec();
    }

    async create(dto: CreateDealerDTO): Promise<DealerDocument> {
        const dealer = await new Dealer(dto).save();
        for (const vehicle of dealer.vehicles) {
            await this.vehicleService.setDealer(vehicle, dealer, false);
        }
        return dealer;
    }

    async update(dto: UpdateDealerDTO): Promise<DealerDocument> {
        if (dto._id == null) {
            throw new MissingKeyInJSONBodyError(['_id']);
        }
        if (dto.vehicles != null) {
            await this.setVehicles(
                new mongoose.Types.ObjectId(dto._id),
                dto.vehicles.map(v => new mongoose.Types.ObjectId(v)),
                true
            );
        }
        delete dto.vehicles;
        const dealer = await Dealer.findOneAndUpdate({ _id: dto._id }, dto, { new: true, runValidators: true }).exec();
        if (dealer == null) {
            throw new EntityDoesNotExistError(`Dealer with id ${dto._id} does not exist.`);
        }
        return dealer;
    }

    async remove(id: string): Promise<DeleteResponse> {
        const dealer: DealerDocument | null = await Dealer.findOne({ _id: id }).exec();
        if (dealer == null) {
            throw new EntityDoesNotExistError(`Dealer with id ${id} does not exist.`);
        }
        if ((dealer.vehicles as mongoose.Types.ObjectId[]).length !== 0) {
            throw new ConstraintError('Dealer has vehicles, delete them first before deleting the dealer.');
        }
        const response = await Dealer.deleteOne({ _id: id }).exec();
        if (response.deletedCount === 0) {
            throw new EntityDoesNotExistError(`Dealer with id ${id} does not exist.`);
        }
        return response;
    }

    async removeVehicle(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehicle: mongoose.Types.ObjectId | VehicleDocument
    ): Promise<void> {
        const dealer = await this.getDealerDocument(dealerOrId);
        dealer.vehicles.pull(vehicle);
        await dealer.save();
    }

    async setVehicles(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehiclesToSet: Array<VehicleDocument | mongoose.Types.ObjectId>,
        updateOppositeSide = false
    ): Promise<void> {
        const dealer: DealerDocument = await this.getDealerDocument(dealerOrId);
        const vehiclesToAdd: Array<mongoose.Types.ObjectId | VehicleDocument> = [];
        // First we check if array contains new vehicles
        for (const vehicleToSet of vehiclesToSet) {
            let hasVehicle = false;
            for (const vehicle of dealer.vehicles) {
                if (vehicleToSet._id.equals(vehicle._id)) {
                    hasVehicle = true;
                    break;
                }
            }
            if (!hasVehicle) {
                vehiclesToAdd.push(vehicleToSet._id);
            }
        }
        // Then we check if any vehicle would be deleted from current list of vehicles
        for (const vehicle of dealer.vehicles) {
            let hasVehicle = false;
            for (const vehicleToSet of vehiclesToSet) {
                if (vehicleToSet._id.equals(vehicle._id)) {
                    hasVehicle = true;
                    break;
                }
            }
            if (!hasVehicle) {
                throw new ConstraintError(`Impossible to delete vehicle ${vehicle._id.toString()} from the dealer, as it wouldn't belong to a dealer anymore.`);
            }
        }
        dealer.vehicles.addToSet(...vehiclesToAdd);
        await dealer.save();
        if (updateOppositeSide) {
            for (const vehicle of vehiclesToAdd) {
                await this.vehicleService.setDealer(vehicle, dealer, false);
            }
        }
    }

    async addVehicles(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehicles: Array<VehicleDocument | mongoose.Types.ObjectId>,
        updateOppositeSide = false
    ): Promise<void> {
        const dealer = await this.getDealerDocument(dealerOrId);
        const ids: mongoose.Types.ObjectId[] = [];
        for (const vehicle of vehicles) {
            if (vehicle instanceof mongoose.Types.ObjectId) {
                ids.push(vehicle);
            } else {
                ids.push(vehicle._id);
            }
        }
        dealer.vehicles.addToSet(ids);
        await dealer.save();
        if (updateOppositeSide) {
            for (const vehicle of vehicles) {
                await this.vehicleService.setDealer(vehicle, dealer, false);
            }
        }
    }

    async addVehicle(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehicle: VehicleDocument | mongoose.Types.ObjectId,
        updateOppositeSide = false
    ): Promise<void> {
        await this.addVehicles(dealerOrId, [vehicle], updateOppositeSide);
    }

    async getDealerDocument(dealerOrDealerId: DealerDocument | mongoose.Types.ObjectId): Promise<DealerDocument> {
        if (dealerOrDealerId instanceof mongoose.Types.ObjectId) {
            const newDealer = await Dealer.findOne({ _id: dealerOrDealerId }).exec();
            if (newDealer == null) {
                throw new EntityDoesNotExistError(`Dealer with id ${dealerOrDealerId.toString()} does not exist.`);
            }
            return newDealer;
        } else {
            return dealerOrDealerId;
        }
    }
}
