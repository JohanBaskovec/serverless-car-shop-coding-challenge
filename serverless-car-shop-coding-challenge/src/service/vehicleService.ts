import { Vehicle } from '../model/vehicle.js';
import { type CreateVehicleDTO, type UpdateVehicleDTO } from '../model/dto/vehicleDTO.js';
import { EntityDoesNotExistError, MissingKeyInJSONBodyError } from '../error/errors.js';
import { type DealerService } from './dealerService.js';
import { type DealerDocument, type VehicleDocument } from '../generated/mongoose.gen.js';
import mongoose from 'mongoose';
import { type CrudService } from './crudService.js';
import { type APIGatewayProxyEventQueryStringParameters } from 'aws-lambda/trigger/api-gateway-proxy.js';
import { type DeleteResponse } from '../model/response.js';

export class VehicleService implements CrudService<VehicleDocument, CreateVehicleDTO, UpdateVehicleDTO> {
    private _dealerService: DealerService;

    get dealerService(): DealerService {
        return this._dealerService;
    }

    set dealerService(value: DealerService) {
        this._dealerService = value;
    }

    async getAll(queryStringParameters: APIGatewayProxyEventQueryStringParameters | null = null): Promise<VehicleDocument[]> {
        const params: any = {};
        if (queryStringParameters != null) {
            if (queryStringParameters.dealer != null) {
                params.dealer = queryStringParameters.dealer;
            }
        }
        return await Vehicle.find(params).exec();
    }

    async create(dto: CreateVehicleDTO): Promise<VehicleDocument> {
        const vehicle = await new Vehicle(dto).save();
        await this.dealerService.addVehicle(vehicle.dealer, vehicle, false);
        return vehicle;
    }

    async update(dto: UpdateVehicleDTO): Promise<VehicleDocument> {
        if (dto._id == null) {
            throw new MissingKeyInJSONBodyError(['_id']);
        }
        const vehicle = await Vehicle.findOne({ _id: dto._id }).exec();
        if (vehicle == null) {
            throw new EntityDoesNotExistError(`Vehicle with id ${dto._id} does not exist.`);
        }
        if (dto.dealer != null) {
            await this.setDealer(vehicle, new mongoose.Types.ObjectId(dto.dealer), true);
        }
        delete dto.dealer;
        const newVehicle = await Vehicle.findOneAndUpdate({ _id: dto._id }, dto, { new: true, runValidators: true }).exec();
        if (newVehicle == null) {
            throw new EntityDoesNotExistError(`Vehicle with id ${dto._id} does not exist.`);
        }
        return newVehicle;
    }

    async remove(id: string): Promise<DeleteResponse> {
        const vehicle = await Vehicle.findOne({ _id: id }).exec();
        if (vehicle == null) {
            throw new EntityDoesNotExistError(`Vehicle with id ${id} does not exist.`);
        }
        await this.dealerService.removeVehicle(vehicle.dealer, vehicle);
        return await Vehicle.deleteOne({ _id: id }).exec();
    }

    private async getVehicleDocument(vehicleOrVehicleId: VehicleDocument | mongoose.Types.ObjectId | string): Promise<VehicleDocument> {
        if (vehicleOrVehicleId instanceof mongoose.Types.ObjectId || typeof vehicleOrVehicleId === 'string') {
            const newVehicle = await Vehicle.findOne({ _id: vehicleOrVehicleId }).exec();
            if (newVehicle == null) {
                throw new EntityDoesNotExistError(`Vehicle with id ${vehicleOrVehicleId.toString()} does not exist.`);
            }
            return newVehicle;
        } else {
            return vehicleOrVehicleId;
        }
    }

    async setDealer(
        vehicleOrId: VehicleDocument | mongoose.Types.ObjectId,
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        updateOppositeSide = false
    ): Promise<void> {
        const vehicle = await this.getVehicleDocument(vehicleOrId);

        await this.dealerService.removeVehicle(vehicle.dealer, vehicle);
        if (updateOppositeSide) {
            await this.dealerService.addVehicle(dealerOrId, vehicle, false);
        }

        vehicle.dealer = dealerOrId;
        await vehicle.save();
    }
}
