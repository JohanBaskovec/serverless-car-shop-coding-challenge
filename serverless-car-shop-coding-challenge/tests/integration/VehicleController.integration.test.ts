import { VehicleService } from '../../src/service/vehicleService.js';
import { Vehicle } from '../../src/model/vehicle.js';
import { Dealer } from '../../src/model/dealer.js';
import { connectToMongo } from '../../src/model/mongoose-db.js';
import { initializeEnvironmentVariables } from '../../src/utils/configuration.js';
import { VehicleController } from '../../src/controller/vehicleController.js';
import { type CreateVehicleDTO, type UpdateVehicleDTO } from '../../src/model/dto/vehicleDTO.js';
import { type VehicleDocument } from '../../src/generated/mongoose.gen.js';
import mongoose from 'mongoose';
import { resetDatabase } from './fakeData.js';
import { assertNotNull } from '../utils/assert.js';
import { DealerService } from '../../src/service/dealerService.js';
import { expectMongoArrayNotToContainId, expectMongoArrayToContainId, getObjectIdAsString } from './utils.js';

beforeAll(async() => {
    initializeEnvironmentVariables();
    await connectToMongo();
});

afterAll(async() => {
    await mongoose.disconnect();
});

function compareVehicles(a: CreateVehicleDTO | VehicleDocument | null, b: CreateVehicleDTO | VehicleDocument | null): void {
    a = assertNotNull(a);
    b = assertNotNull(b);
    expect(a.make).toEqual(b.make);
    expect(a.color).toEqual(b.color);
    assertNotNull(a.dealer);
    assertNotNull(b.dealer);
    expect(getObjectIdAsString(a.dealer)).toEqual(getObjectIdAsString(b.dealer));
    expect(a.fuelType).toEqual(b.fuelType);
    expect(a.mileage).toEqual(b.mileage);
    expect(a.vehicleModel).toEqual(b.vehicleModel);
    expect(a.transmission).toEqual(b.transmission);
    expect(a.vehicleType).toEqual(b.vehicleType);
}

const dealerService = new DealerService();
const vehicleService = new VehicleService();
dealerService.vehicleService = vehicleService;
vehicleService.dealerService = dealerService;
const vehicleController = new VehicleController(vehicleService);

beforeEach(async() => {
    await resetDatabase();
});

describe('VehicleController', () => {
    describe('getAllVehicles()', () => {
        test('should return an empty array if there are no vehicles', async() => {
            await Vehicle.deleteMany({}).exec();
            const response = await vehicleController.getAll();
            expect(response).toMatchSnapshot();
        });

        test('should return vehicles when there are vehicles', async() => {
            const response = await vehicleController.getAll();
            expect(response).toMatchSnapshot();
        });

        test('should return vehicles when filtering by dealer', async() => {
            const dealer = assertNotNull(await Dealer.findOne({ _id: '644171318ce5b4f5098621cb' }).exec());
            const response = await vehicleController.getAll({ dealer: dealer._id.toString() });
            expect(response).toMatchSnapshot();
        });
    });

    describe('createVehicle()', () => {
        test('should create a vehicle when the request is correct', async() => {
            const dealerBefore = assertNotNull(await Dealer.findOne({}).exec());
            const dto: CreateVehicleDTO = {
                make: 'make2',
                color: 'orange',
                dealer: dealerBefore.id,
                fuelType: 'lpg',
                mileage: 4,
                vehicleModel: 'model3',
                transmission: 'automatic',
                vehicleType: 'other'
            };
            const timestampBeforeRequest = +Date.now();
            const response = await vehicleController.create(JSON.stringify(dto));
            const timeStampAfterRequest = +Date.now();

            const vehicleResponse = JSON.parse(response.body).data;
            compareVehicles(vehicleResponse, dto);
            const creationDateTimestamp = +new Date(vehicleResponse.creationDate);
            const lastUpdateTimestamp = +new Date(vehicleResponse.lastUpdateDate);
            expect(creationDateTimestamp).toBeGreaterThanOrEqual(timestampBeforeRequest);
            expect(creationDateTimestamp).toBeLessThanOrEqual(timeStampAfterRequest);
            expect(lastUpdateTimestamp).toBeGreaterThanOrEqual(timestampBeforeRequest);
            expect(lastUpdateTimestamp).toBeLessThanOrEqual(timeStampAfterRequest);

            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleResponse._id }).exec());
            compareVehicles(dto, vehicleInDb);

            const dealerAfter = assertNotNull(await Dealer.findOne({ _id: vehicleResponse.dealer }).exec());
            expectMongoArrayToContainId(dealerAfter.vehicles, vehicleInDb);
        });

        test('should return an error when no body provided', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const response = await vehicleController.create(null);
            expect(response).toMatchSnapshot();

            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            expect(vehicleInDb.vehicleModel).toBe(vehicleToUpdate.vehicleModel);
        });

        test('should return an error when there are missing fields', async() => {
            const dto: any = {};
            const response = await vehicleController.create(JSON.stringify(dto));
            expect(response).toMatchSnapshot();
        });

        test('should return an error when trying to use an invalid value', async() => {
            const dealer = assertNotNull(await Dealer.findOne({}).exec());
            const dto: any = {
                make: 'make2',
                color: 'orange',
                dealer: dealer.id,
                fuelType: 'invalid-fuel-type',
                mileage: 4,
                vehicleModel: 'model3',
                transmission: 'automatic',
                vehicleType: 'other'
            };
            const response = await vehicleController.create(JSON.stringify(dto));
            expect(response).toMatchSnapshot();
        });

        test('should return an error when using an invalid dealer id', async() => {
            const dto: any = {
                make: 'make2',
                color: 'orange',
                dealer: 'not-an-id',
                fuelType: 'lpg',
                mileage: 4,
                vehicleModel: 'model3',
                transmission: 'automatic',
                vehicleType: 'other'
            };
            const response = await vehicleController.create(JSON.stringify(dto));
            expect(response).toMatchSnapshot();
        });
    });

    describe('removeVehicle()', () => {
        test('should remove the vehicle when the request is correct', async() => {
            const vehicleToRemove: VehicleDocument = assertNotNull(await Vehicle.findOne({}).populate('dealer').exec());
            const nVehiclesBefore = await Vehicle.count().exec();
            const response = await vehicleController.remove({ id: vehicleToRemove._id.toString() });
            const nVehiclesAfter = await Vehicle.count().exec();
            expect(response).toMatchSnapshot();

            const vehicleInDb = await Vehicle.findOne({ _id: vehicleToRemove._id }).exec();
            expect(vehicleInDb).toBeNull();
            expect(nVehiclesBefore).toBe(nVehiclesAfter + 1);

            const dealerAfter = assertNotNull(await Dealer.findOne({ _id: vehicleToRemove.dealer._id }).exec());
            expectMongoArrayNotToContainId(dealerAfter.vehicles, vehicleToRemove);
        });

        test('should return an error when no id provided', async() => {
            const nVehiclesBefore = await Vehicle.count().exec();
            const response = await vehicleController.remove({});
            const nVehiclesAfter = await Vehicle.count().exec();
            expect(response).toMatchSnapshot();
            expect(nVehiclesBefore).toBe(nVehiclesAfter);
        });

        test('should return an error when vehicle with id does not exist', async() => {
            const id = '754171318ce5b4f5098621cb';
            const nVehiclesBefore = await Vehicle.count().exec();
            const response = await vehicleController.remove({ id });
            const nVehiclesAfter = await Vehicle.count().exec();
            expect(response).toMatchSnapshot();
            expect(nVehiclesBefore).toBe(nVehiclesAfter);
        });
    });

    describe('updateVehicle()', () => {
        test('should update the vehicle when the request is correct', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const newDealerId = '754171318ce5b4f5098621cb';
            const newDealer = await new Dealer({
                _id: newDealerId,
                name: 'Dealer 3'
            }).save();

            const dto: UpdateVehicleDTO = {
                _id: vehicleToUpdate._id.toString(),
                vehicleModel: 'super model!',
                dealer: newDealerId
            };
            const timestampBeforeRequest = +Date.now();
            const response = await vehicleController.update(JSON.stringify(dto));
            const timeStampAfterRequest = +Date.now();
            const vehicleResponse = JSON.parse(response.body).data;
            expect(response.statusCode).toBe(200);

            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            compareVehicles(vehicleInDb, vehicleResponse);
            const lastUpdateTimestamp = +assertNotNull(vehicleInDb.lastUpdateDate);
            expect(lastUpdateTimestamp).toBeGreaterThanOrEqual(timestampBeforeRequest);
            expect(lastUpdateTimestamp).toBeLessThanOrEqual(timeStampAfterRequest);
            expect(vehicleInDb.vehicleModel).toBe(dto.vehicleModel);
            expect(getObjectIdAsString(vehicleInDb.dealer)).toBe(newDealerId);

            const dealerAfter = assertNotNull(await Dealer.findOne({ _id: newDealer.id }).exec());
            expectMongoArrayToContainId(dealerAfter.vehicles, vehicleInDb);

            const previousDealer = assertNotNull(await Dealer.findOne({ _id: vehicleToUpdate.dealer.id }).exec());
            expectMongoArrayNotToContainId(previousDealer.vehicles, vehicleInDb);
        });

        test('should return an error when no body provided', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const response = await vehicleController.update(null);
            expect(response).toMatchSnapshot();

            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            expect(vehicleInDb.vehicleModel).toBe(vehicleToUpdate.vehicleModel);
        });

        test('should return an error when no id provided', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const dto: any = {
                vehicleModel: 'super model!'
            };
            const response = await vehicleController.update(JSON.stringify(dto));
            expect(response).toMatchSnapshot();

            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            expect(vehicleInDb.vehicleModel).toBe(vehicleToUpdate.vehicleModel);
        });

        test('should return an error when vehicle with id does not exist', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const dto: UpdateVehicleDTO = {
                _id: '754171318ce5b4f5098621cb',
                vehicleModel: 'super model!'
            };
            const response = await vehicleController.update(JSON.stringify(dto));
            expect(response).toMatchSnapshot();

            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            expect(vehicleInDb.vehicleModel).toBe(vehicleToUpdate.vehicleModel);
        });

        test('should return an error when using invalid fuel type', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const response = await vehicleController.update(JSON.stringify({
                _id: vehicleToUpdate._id.toString(),
                fuelType: 'invalid fuel type'
            }));
            expect(response).toMatchSnapshot();
            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            expect(vehicleInDb.fuelType).toBe(vehicleToUpdate.fuelType);
        });

        test('should return an error when using invalid transmission', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const response = await vehicleController.update(JSON.stringify({
                _id: vehicleToUpdate._id.toString(),
                transmission: 'invalid transmission'
            }));
            expect(response).toMatchSnapshot();
            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            expect(vehicleInDb.transmission).toBe(vehicleToUpdate.transmission);
        });

        test('should return an error when using mileage of wrong type', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const response = await vehicleController.update(JSON.stringify({
                _id: vehicleToUpdate._id.toString(),
                mileage: 'invalid mileage'
            }));
            expect(response).toMatchSnapshot();
            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            expect(vehicleInDb.mileage).toBe(vehicleToUpdate.mileage);
        });

        test('should return an error with empty id', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const response = await vehicleController.update(JSON.stringify({
                _id: '',
                make: 'new make'
            }));
            expect(response).toMatchSnapshot();
            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            expect(vehicleInDb.mileage).toBe(vehicleToUpdate.mileage);
        });

        test('should return an error with empty dealer id', async() => {
            const vehicleToUpdate = assertNotNull(await Vehicle.findOne({}).exec());
            const response = await vehicleController.update(JSON.stringify({
                _id: vehicleToUpdate._id.toString(),
                dealer: ''
            }));
            expect(response).toMatchSnapshot();
            const vehicleInDb = assertNotNull(await Vehicle.findOne({ _id: vehicleToUpdate._id }).exec());
            expect(vehicleInDb.mileage).toBe(vehicleToUpdate.mileage);
        });
    });
});
