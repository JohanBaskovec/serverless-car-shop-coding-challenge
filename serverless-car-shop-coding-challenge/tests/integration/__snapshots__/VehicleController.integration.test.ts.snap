// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleController createVehicle() should return an error when no body provided 1`] = `
{
  "body": "{"error":{"type":"MissingRequestBody","message":"Missing body in HTTP request."}}",
  "statusCode": 400,
}
`;

exports[`VehicleController createVehicle() should return an error when there are missing fields 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Vehicle validation failed: dealer: Path \`dealer\` is required., color: Path \`color\` is required., vehicleType: Path \`vehicleType\` is required., mileage: Path \`mileage\` is required., fuelType: Path \`fuelType\` is required., transmission: Path \`transmission\` is required., vehicleModel: Path \`vehicleModel\` is required., make: Path \`make\` is required."}}",
  "statusCode": 400,
}
`;

exports[`VehicleController createVehicle() should return an error when trying to use an invalid value 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Vehicle validation failed: fuelType: \`invalid-fuel-type\` is not a valid enum value for path \`fuelType\`."}}",
  "statusCode": 400,
}
`;

exports[`VehicleController createVehicle() should return an error when using an invalid dealer id 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Vehicle validation failed: dealer: Cast to ObjectId failed for value \\"not-an-id\\" (type string) at path \\"dealer\\" because of \\"BSONError\\""}}",
  "statusCode": 400,
}
`;

exports[`VehicleController getAllVehicles() should return an empty array if there are no vehicles 1`] = `
{
  "body": "{"data":[]}",
  "statusCode": 200,
}
`;

exports[`VehicleController getAllVehicles() should return vehicles when filtering by dealer 1`] = `
{
  "body": "{"data":[{"_id":"644170e85d3ab81e12fcca7c","make":"make1","vehicleModel":"model1","transmission":"manual","fuelType":"lpg","mileage":0,"vehicleType":"coupe","color":"blue","dealer":"644171318ce5b4f5098621cb","creationDate":"2010-10-10T00:00:00.000Z","lastUpdateDate":"2010-10-10T00:00:00.000Z","__v":0},{"_id":"644170e85d3ab81e12fcca7e","make":"make2","vehicleModel":"model2","transmission":"automatic","fuelType":"lpg","mileage":10,"vehicleType":"van","color":"red","dealer":"644171318ce5b4f5098621cb","creationDate":"2012-10-10T00:00:00.000Z","lastUpdateDate":"2012-10-10T00:00:00.000Z","__v":0}]}",
  "statusCode": 200,
}
`;

exports[`VehicleController getAllVehicles() should return vehicles when there are vehicles 1`] = `
{
  "body": "{"data":[{"_id":"644170e85d3ab81e12fcca7c","make":"make1","vehicleModel":"model1","transmission":"manual","fuelType":"lpg","mileage":0,"vehicleType":"coupe","color":"blue","dealer":"644171318ce5b4f5098621cb","creationDate":"2010-10-10T00:00:00.000Z","lastUpdateDate":"2010-10-10T00:00:00.000Z","__v":0},{"_id":"644170e85d3ab81e12fcca7e","make":"make2","vehicleModel":"model2","transmission":"automatic","fuelType":"lpg","mileage":10,"vehicleType":"van","color":"red","dealer":"644171318ce5b4f5098621cb","creationDate":"2012-10-10T00:00:00.000Z","lastUpdateDate":"2012-10-10T00:00:00.000Z","__v":0},{"_id":"644170e85d3ab81e12fcca7f","make":"make2","vehicleModel":"model2","transmission":"automatic","fuelType":"lpg","mileage":10,"vehicleType":"van","color":"red","dealer":"644171318ce5b4f5098621ca","creationDate":"2011-10-10T00:00:00.000Z","lastUpdateDate":"2011-10-10T00:00:00.000Z","__v":0}]}",
  "statusCode": 200,
}
`;

exports[`VehicleController removeVehicle() should remove the vehicle when the request is correct 1`] = `
{
  "body": "{"data":{"acknowledged":true,"deletedCount":1}}",
  "statusCode": 200,
}
`;

exports[`VehicleController removeVehicle() should return an error when no id provided 1`] = `
{
  "body": "{"error":{"type":"MissingPathParametersError","message":"Missing path parameters: id"}}",
  "statusCode": 400,
}
`;

exports[`VehicleController removeVehicle() should return an error when vehicle with id does not exist 1`] = `
{
  "body": "{"error":{"type":"EntityDoesNotExistError","message":"Vehicle with id 754171318ce5b4f5098621cb does not exist."}}",
  "statusCode": 404,
}
`;

exports[`VehicleController updateVehicle() should return an error when no body provided 1`] = `
{
  "body": "{"error":{"type":"MissingRequestBody","message":"Missing body in HTTP request."}}",
  "statusCode": 400,
}
`;

exports[`VehicleController updateVehicle() should return an error when no id provided 1`] = `
{
  "body": "{"error":{"type":"MissingKeyInJSONBodyError","message":"Missing key in JSON: _id"}}",
  "statusCode": 400,
}
`;

exports[`VehicleController updateVehicle() should return an error when using invalid fuel type 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Validation failed: fuelType: \`invalid fuel type\` is not a valid enum value for path \`fuelType\`."}}",
  "statusCode": 400,
}
`;

exports[`VehicleController updateVehicle() should return an error when using invalid transmission 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Validation failed: transmission: \`invalid transmission\` is not a valid enum value for path \`transmission\`."}}",
  "statusCode": 400,
}
`;

exports[`VehicleController updateVehicle() should return an error when using mileage of wrong type 1`] = `
{
  "body": "{"error":{"type":"CastError","message":"Cast to Number failed for value \\"invalid mileage\\" (type string) at path \\"mileage\\""}}",
  "statusCode": 400,
}
`;

exports[`VehicleController updateVehicle() should return an error when vehicle with id does not exist 1`] = `
{
  "body": "{"error":{"type":"EntityDoesNotExistError","message":"Vehicle with id 754171318ce5b4f5098621cb does not exist."}}",
  "statusCode": 404,
}
`;

exports[`VehicleController updateVehicle() should return an error with empty dealer id 1`] = `
{
  "body": "{"error":{"type":"BSONError","message":"Argument passed in must be a string of 12 bytes or a string of 24 hex characters or an integer"}}",
  "statusCode": 400,
}
`;

exports[`VehicleController updateVehicle() should return an error with empty id 1`] = `
{
  "body": "{"error":{"type":"CastError","message":"Cast to ObjectId failed for value \\"\\" (type string) at path \\"_id\\" for model \\"Vehicle\\""}}",
  "statusCode": 400,
}
`;
